In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [ ]:
dataframe = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/ML Concordia/Major Assignment 2/archive/survey lung cancer.csv")

In [ ]:
dataframe_num = dataframe
replacement_map = {'M': 1, 'F': 0}
dataframe_num['GENDER'] = dataframe_num['GENDER'].replace(replacement_map)
dataframe_num = dataframe
replacement_map = {'YES': 1, 'NO': 0}
dataframe_num['LUNG_CANCER'] = dataframe_num['LUNG_CANCER'].replace(replacement_map)

In [ ]:
dataframe_x = dataframe_num[dataframe_num.columns.difference(['LUNG_CANCER'])]
X_train, X_test, y_train, y_test = train_test_split(dataframe_x, dataframe_num['LUNG_CANCER'], test_size=0.3)

In [ ]:
#K-nearest
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=10)
knn_model.fit(X_train, y_train)

train_predictions = knn_model.predict(X_train)
test_predictions = knn_model.predict(X_test)

print("Classification Report for Training Set:")
print(classification_report(y_train, train_predictions))

print("\nClassification Report for Testing Set:")
print(classification_report(y_test, test_predictions))

Classification Report for Training Set:
              precision    recall  f1-score   support

           0       0.50      0.04      0.08        23
           1       0.90      0.99      0.94       193

    accuracy                           0.89       216
   macro avg       0.70      0.52      0.51       216
weighted avg       0.85      0.89      0.85       216


Classification Report for Testing Set:
              precision    recall  f1-score   support

           0       0.67      0.12      0.21        16
           1       0.84      0.99      0.91        77

    accuracy                           0.84        93
   macro avg       0.76      0.56      0.56        93
weighted avg       0.81      0.84      0.79        93



In [ ]:
#SVM
from sklearn.svm import SVC

svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)

train_predictions = svm_model.predict(X_train)
test_predictions = svm_model.predict(X_test)

print("Classification Report for Training Set:")
print(classification_report(y_train, train_predictions))

print("\nClassification Report for Testing Set:")
print(classification_report(y_test, test_predictions))

Classification Report for Training Set:
              precision    recall  f1-score   support

           0       0.75      0.78      0.77        23
           1       0.97      0.97      0.97       193

    accuracy                           0.95       216
   macro avg       0.86      0.88      0.87       216
weighted avg       0.95      0.95      0.95       216


Classification Report for Testing Set:
              precision    recall  f1-score   support

           0       0.75      0.56      0.64        16
           1       0.91      0.96      0.94        77

    accuracy                           0.89        93
   macro avg       0.83      0.76      0.79        93
weighted avg       0.89      0.89      0.89        93



In [ ]:
# Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB

gnb_model = GaussianNB()
gnb_model.fit(X_train, y_train)

train_predictions = gnb_model.predict(X_train)
test_predictions = gnb_model.predict(X_test)

print("Classification Report for Training Set:")
print(classification_report(y_train, train_predictions))

print("\nClassification Report for Testing Set:")
print(classification_report(y_test, test_predictions))

Classification Report for Training Set:
              precision    recall  f1-score   support

           0       0.58      0.65      0.61        23
           1       0.96      0.94      0.95       193

    accuracy                           0.91       216
   macro avg       0.77      0.80      0.78       216
weighted avg       0.92      0.91      0.91       216


Classification Report for Testing Set:
              precision    recall  f1-score   support

           0       0.75      0.56      0.64        16
           1       0.91      0.96      0.94        77

    accuracy                           0.89        93
   macro avg       0.83      0.76      0.79        93
weighted avg       0.89      0.89      0.89        93



In [ ]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier

tree_model = DecisionTreeClassifier()
tree_model.fit(X_train, y_train)

train_predictions = tree_model.predict(X_train)
test_predictions = tree_model.predict(X_test)

print("Classification Report for Training Set:")
print(classification_report(y_train, train_predictions))

print("\nClassification Report for Testing Set:")
print(classification_report(y_test, test_predictions))

Classification Report for Training Set:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        23
           1       1.00      1.00      1.00       193

    accuracy                           1.00       216
   macro avg       1.00      1.00      1.00       216
weighted avg       1.00      1.00      1.00       216


Classification Report for Testing Set:
              precision    recall  f1-score   support

           0       0.62      0.31      0.42        16
           1       0.87      0.96      0.91        77

    accuracy                           0.85        93
   macro avg       0.75      0.64      0.67        93
weighted avg       0.83      0.85      0.83        93



In [ ]:
#AdaBoost
from sklearn.ensemble import AdaBoostClassifier

adaboost_model = AdaBoostClassifier()
adaboost_model.fit(X_train, y_train)

train_predictions = adaboost_model.predict(X_train)
test_predictions = adaboost_model.predict(X_test)

print("Classification Report for Training Set:")
print(classification_report(y_train, train_predictions))

print("\nClassification Report for Testing Set:")
print(classification_report(y_test, test_predictions))

print('''By default, the AdaBoostClassifier utilizes a decision tree with a maximum depth of 1 (max_depth=1), commonly referred to as a decision stump, as its base estimator.
This decision stump represents a simple tree model comprising a single decision node and two leaf nodes''')

Classification Report for Training Set:
              precision    recall  f1-score   support

           0       0.86      0.78      0.82        23
           1       0.97      0.98      0.98       193

    accuracy                           0.96       216
   macro avg       0.92      0.88      0.90       216
weighted avg       0.96      0.96      0.96       216


Classification Report for Testing Set:
              precision    recall  f1-score   support

           0       0.62      0.31      0.42        16
           1       0.87      0.96      0.91        77

    accuracy                           0.85        93
   macro avg       0.75      0.64      0.67        93
weighted avg       0.83      0.85      0.83        93

By default, the AdaBoostClassifier utilizes a decision tree with a maximum depth of 1 (max_depth=1), commonly referred to as a decision stump, as its base estimator. 
This decision stump represents a simple tree model comprising a single decision node and two leaf n

#Discuss and compare the performance of the trained models.
The K-nearest model with K=10 has the worst performance, especially in predicting no-cancer dataset. As the report shows, the accuracy of the model is around 50% while the accuracy for predicting cancer is over 90%. This bad forecast does not have much effect on the final accuracy because the number of healthy items is very small. As a result, the report shows an accuracy of prediction around 80-90%.

The SVM and GaussianNB are similar in the accuracy of predictions. They have a high accuracy in training and testing sets. To explain more, SVM has a high accuracy in predicting cancer and acceptable accuracy in predicting no-cancer.  GaussianNB has a similarly high accuracy in the prediction of cancer. While the prediction of no-cancer is less than the same situation in SVM. Finally, the accuracy of both of them is around 90%.

Overfitting occurred in the decision tree. The decision tree correctly predicts all data sets in training, while the prediction accuracy is lower in the testing set. The precision of prediction in the testing set is around 60% in no-cancer and more than 90% in cancer situations. Finally, the accuracy of the testing set is 85%.

AdaBoost predicts cancer and no-cancer of training datasets with a precision of 80-90%, in both situations. The accuracy of prediction of the training dataset is 96%. While the precision for testing dataset is a little different. The precision of predicting cancer and no cancer are around 87% and 62%, respectively. Finally, the accuracy of prediction of testing data is 85%.